### DISCUSSION TITLE
Ability to use Container outside of request handling

### DISCUSSION THREAD

**Discussion by hvt** [clue/framework-x#243] [Q&A] [answered]
Hey there,

I'm thinkering around with Framework X, and still like many of the details.

For my test project, I'm also in need of a 'command'. I.e. a task that is ran every once in a while. I know there are plans to support commands as part of the framework, but to get ahead of that, I thought to re-use some of the services I defined in the `FrameworkX\Container`.

However, I am unable to use the container outside of a request handling perspective. It looks to me that the container is only to be used as a middleware (the `__invoke()` implementation). Is this assumption correct?

Is there any way to retrieve services from the container outside of request handling perspective?

Besides this question, what about the idea to let the container itself also implement PSR-11?

**Comment by SimonFrings** [comment, marked as answer]
Hey @hvt, thanks for bringing this up, glad you like Framework X 👍️

> For my test project, I'm also in need of a 'command'. I.e. a task that is ran every once in a while

Definitely a great feature request, sounds like you want to use something like CRON which is one of the features we also talked about while working on the recently released `v0.16.0`. I know that @clue has a prototype for this on his laptop and we also see this as a great feature addition that we want to introduce not too far in the future. We're currently heading towards a `v0.17.0` which will improve/rework some of the handler structure, so we can then go ahead and add a few new handlers too Framework X (won't spoil too much 🤫). If you need this more urgently, you can also reach out to us or help us prioritizing by [becoming a sponsor](https://github.com/sponsors/clue) ❤️

> For my test project, I'm also in need of a 'command'. I.e. a task that is ran every once in a while. I know there are plans to support commands as part of the framework, but to get ahead of that, I thought to re-use some of the services I defined in the FrameworkX\Container.
>
> However, I am unable to use the container outside of a request handling perspective. It looks to me that the container is only to be used as a middleware (the __invoke() implementation). Is this assumption correct?
>
> Is there any way to retrieve services from the container outside of request handling perspective?
>
> Besides this question, what about the idea to let the container itself also implement PSR-11?

The `Container` is not really designed for this as we don't want to overload this with complexity and mostly use it for dependency injection. Additionally, we already support the PSR-11 container interface to pass in custom containers, but we're not sure if we should also support PSR-11 OOTB, you can read more about this in our [PSR-11 discussion](https://github.com/clue/framework-x/discussions/159).

Hope this helps and gives you some insights into our plans for Framework X :+1:
